### PR TITLE
Bump SQLAlchemy requirement to >1.4.0

### PR DESCRIPTION
**Pull request by tdesveaux**
Draft until warn when using different DBs are resolved.

Relates to #7591.

This change the minimum required version of SQLAlchemy to 1.4.0 which is a version made to prepare upgrade to 2.0.
Also bumps SQLAlchemy for CI to 1.4.52.

Also fixes the majority of `RemovedIn20Warning` warnings, leaving 3 unresolved ones as their fixes are less simple than the others.
Each warning fix is a self contained commit to ease review, but I can split this in mulitple PRs if needed.

Maybe `SQLALCHEMY_WARN_20=1` should be set in env in `master/buildbot/test/__init__.py` as well, otherwise, running trial locally without it will result in errors as the warn message is a generic one:
`Deprecated API features detected! These feature(s) are not compatible with SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQLALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message.`

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `newsfragments` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/db/dbconfig.py [hunk 1]
```diff
@@ -61,7 +61,8 @@ def getDb(self):
         db.model = model.Model(db)
         db.state = state.StateConnectorComponent(db)
         try:
-            self.objectid = db.state.thdGetObjectId(db_engine, self.name, "DbConfig")['id']
+            with db_engine.connect() as conn:
```
Comment (p12tic) [inline master/buildbot/db/dbconfig.py:64]: The thread pool is to ensure that the main thread does not block. Is db_engine.connect() guaranteed not to block even for a short time?

File: master/buildbot/db/logs.py [hunk 2]
```diff
@@ -427,7 +427,7 @@ def thddeleteOldLogs(conn):
 
         return self.db.pool.do(thddeleteOldLogs)
 
-    def _logdictFromRow(self, row):
-        rv = dict(row)
+    def _logdictFromRow(self, row: sa.engine.row.Row):
+        rv = dict(row._mapping)  # noqa pylint: disable=protected-access
```
Comment (p12tic) [inline master/buildbot/db/logs.py:434]: Using internal variable doesn't seem like a long-term solution. I think we should use mappings() function https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Result.mappings

**Review by p12tic — CHANGES_REQUESTED** [review]
Thanks a lot. Several comments added, should be easy to address.

File: master/buildbot/db/logs.py (hunk 2 above)
Comment (tdesveaux) [reply master/buildbot/db/logs.py:434]: This is indeed far better,
I guess I use the `_mapping` as it lead to less change, but that's a poor reason and `mappings()` is a far better alternative.
I'll update the PR.

File: master/buildbot/db/dbconfig.py (hunk 1 above)
Comment (tdesveaux) [reply master/buildbot/db/dbconfig.py:64]: I think this does not change the blocking aspect of this code as `thdGetObjectId` is the "implementation` with `_getObjectId` being the wrapper that will thread it.

I though direct call to `thdGetObjectId` would block the thread as well, am I wrong?

Comment (p12tic) [reply master/buildbot/db/dbconfig.py:64]: OK, I was confused about which code path this is used in. All good, thanks.

**Pull request merged**